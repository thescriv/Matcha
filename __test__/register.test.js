const crypto = require("crypto")
const superagent = require("superagent")
const dotenv = require("dotenv")

const db = require("../src/lib/db")

const { script } = require("../scripts/script")
const { deleteRows } = require("./utils/deleteRows")
const { transporter } = require("../src/lib/transporter")
const { getRandomPort } = require("./utils/getRandomPort")
const { startApi, stopApi } = require("../api")

let baseUrl

describe(`register --`, () => {
  beforeAll(async () => {
    const port = getRandomPort()

    dotenv.config({ path: "../.env" })

    baseUrl = `http://localhost:${port}/api`

    await script("matchaTest")

    await deleteRows()

    await startApi(port)
  })

  beforeEach(() => {
    jest.spyOn(Date, "now").mockReturnValue(1000000000000)
  })

  afterEach(async () => {
    jest.restoreAllMocks()

    await deleteRows()
  })

  afterAll(async () => {
    await stopApi()
  })

  describe("POST /register", () => {
    test("do register", async () => {
      jest.spyOn(transporter, "sendMail").mockReturnValue(true)
      jest
        .spyOn(crypto, "randomBytes")
        .mockReturnValue("0000000000000000000000000000000000000001")

      const { body, status } = await superagent
        .post(`${baseUrl}/register`)
        .send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [tokenLink] = await db.query("SELECT * FROM register_token")
      expect(tokenLink).toMatchSnapshot()
    })

    test("do not register (query fail on insert but errno != 1062)", async () => {
      jest.spyOn(db, "query").mockImplementation(() => {
        throw new Error("AN ERROR OCCURED")
      })

      let error
      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err.response
      }
      const { body, status } = error

      expect({ body, status }).toMatchSnapshot()
    })

    test("do not register (sendMail failed)", async () => {
      jest.spyOn(transporter, "sendMail").mockReturnValue(false)

      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (email already exist)", async () => {
      await db.query(
        `INSERT INTO user (nickname, firstname, lastname, email, password) 
        VALUES ("thescriv", "theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv1",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (nickname already exist)", async () => {
      await db.query(
        `INSERT INTO user (nickname, firstname, lastname, email, password) 
        VALUES ("thescriv", "theo", "test", "test@test1.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (bad email)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (bad password)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "test",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (no firstname)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: null,
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (no lastname)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: null,
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (no nickname)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: null,
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (no email)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: null,
          password: "testTest1",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })

    test("do not register (no password)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/register`).send({
          nickname: "thescriv",
          firstname: "theo",
          lastname: "test",
          email: "test@test.test",
          password: null,
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()
    })
  })

  describe("GET /register/verify-email/:id", () => {
    test("do verifyEmail", async () => {
      jest.spyOn(transporter, "sendMail").mockReturnValue(true)
      jest
        .spyOn(crypto, "randomBytes")
        .mockReturnValue("0000000000000000000000000000000000000001")

      const [{ insertId }] = await db.query(
        `INSERT INTO user (firstname, lastname, email, password) 
          VALUES ("theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      await db.query(`INSERT INTO register_token (token, user_id) 
        VALUES ("0000000000000000000000000000000000000001", "${insertId}")`)

      const { body, status } = await superagent.get(
        `${baseUrl}/register/verify-email/0000000000000000000000000000000000000001`
      )

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [tokenLink] = await db.query("SELECT * FROM register_token")
      expect(tokenLink).toMatchSnapshot()
    })

    test("do not verifyEmail (id is null)", async () => {
      let error

      try {
        await superagent.get(`${baseUrl}/register/verify-email/null`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })

    test("do not verifyEmail (no id)", async () => {
      let error

      try {
        await superagent.get(`${baseUrl}/register/verify-email/`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })

    test("do not verifyEmail (id does not exist)", async () => {
      let error

      try {
        await superagent.get(
          `${baseUrl}/register/verify-email/0000000000000000000000000000000000000001`
        )
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })
  })
})
