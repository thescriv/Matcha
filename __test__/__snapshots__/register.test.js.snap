// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register -- GET /register/verify-email/:id do not verifyEmail (id does not exist) 1`] = `
Object {
  "body": Object {
    "error": "api.verifyEmail tokenId does_not_exist",
  },
  "status": 400,
}
`;

exports[`register -- GET /register/verify-email/:id do not verifyEmail (id is null) 1`] = `
Object {
  "body": Object {
    "error": "api.verifyEmail tokenId does_not_exist",
  },
  "status": 400,
}
`;

exports[`register -- GET /register/verify-email/:id do not verifyEmail (no id) 1`] = `
Object {
  "body": Object {},
  "status": 404,
}
`;

exports[`register -- GET /register/verify-email/:id do verifyEmail 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`register -- GET /register/verify-email/:id do verifyEmail 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": null,
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 1,
  },
]
`;

exports[`register -- GET /register/verify-email/:id do verifyEmail 3`] = `Array []`;

exports[`register -- POST /register do not register (bad email) 1`] = `
Object {
  "body": Object {
    "error": "api.register email is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (bad email) 2`] = `Array []`;

exports[`register -- POST /register do not register (bad password) 1`] = `
Object {
  "body": Object {
    "error": "api.register password is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (bad password) 2`] = `Array []`;

exports[`register -- POST /register do not register (email already exist) 1`] = `
Object {
  "body": Object {
    "error": "api.register  'user.email' already_exist",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (email already exist) 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`register -- POST /register do not register (nickname already exist) 1`] = `
Object {
  "body": Object {
    "error": "api.register  'user.nickname' already_exist",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (nickname already exist) 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 0,
    "email": "test@test1.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`register -- POST /register do not register (no email) 1`] = `
Object {
  "body": Object {
    "error": "api.register email is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (no email) 2`] = `Array []`;

exports[`register -- POST /register do not register (no firstname) 1`] = `
Object {
  "body": Object {
    "error": "api.register firstname is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (no firstname) 2`] = `Array []`;

exports[`register -- POST /register do not register (no lastname) 1`] = `
Object {
  "body": Object {
    "error": "api.register lastname is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (no lastname) 2`] = `Array []`;

exports[`register -- POST /register do not register (no nickname) 1`] = `
Object {
  "body": Object {
    "error": "api.register nickname is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (no nickname) 2`] = `Array []`;

exports[`register -- POST /register do not register (no password) 1`] = `
Object {
  "body": Object {
    "error": "api.register password is_not_valid",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (no password) 2`] = `Array []`;

exports[`register -- POST /register do not register (query fail on insert but errno != 1062) 1`] = `
Object {
  "body": Object {
    "error": "Error: AN ERROR OCCURED",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (sendMail failed) 1`] = `
Object {
  "body": Object {
    "error": "api.register sendMail failed",
  },
  "status": 400,
}
`;

exports[`register -- POST /register do not register (sendMail failed) 2`] = `Array []`;

exports[`register -- POST /register do register 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`register -- POST /register do register 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`register -- POST /register do register 3`] = `
Array [
  RowDataPacket {
    "creation_date": 1000000000000,
    "id": 1,
    "token": "0000000000000000000000000000000000000001",
    "user_id": 1,
  },
]
`;
